test/pylib/coverage_utils.py

```diff
@@ -0,0 +1,1293 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2015-present ScyllaDB
+#
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+import argparse
+import asyncio
+import multiprocessing
+from pathlib import Path,PurePath
+from typing import Union,List,Dict,Iterable,Optional,Coroutine,Mapping,Any,List,Self
+import re
+from enum import Enum
+from asyncio.subprocess import PIPE,DEVNULL
+from collections.abc import Iterable as IterableType
+import os
+from collections import namedtuple
+from itertools import repeat
+from functools import wraps,partial
+import logging
+import sys
+import inspect
+#So the module can be imported outside of this directory
+sys.path.insert(0,os.path.join(os.path.dirname(__file__)))
+import lcov_utils
+del sys.path[0]
+import concurrent.futures
+from urllib.parse import quote,unquote
+# NOTE: A lot of the functions in this file uses the form: func(*,param1,param2....)
+# This was intentionally done in order to prevent calling those function with positional
+# arguments until the API stabilizes.
+
+# Type definitions for annotation
+PathLike = Union[Path,str]
+PathLikeCollection = Iterable[PathLike]
+ConcurrencyParam = Optional[Union[int,asyncio.Semaphore]]
+PathCollection = Iterable[Path]
+PathLikeOrPathLikeCollection = Union[PathLike,Iterable[PathLike]]
+LoggerType = Union[logging.Logger,logging.LoggerAdapter]
+
+# Logging support for debugging
+COVERAGE_TOOLS_LOGGER = logging.getLogger("coverage_utils.py")
+COVERAGE_TOOLS_LOGGER.addHandler(logging.NullHandler())
+COVERAGE_TOOLS_LOGGER.propagate = False
+
+# Utility functions for tracing
+_tid = 0
+def unique_trace_id():
+    global _tid
+    _tid += 1
+    return _tid
+
+def trace_function_call(func, sig : inspect.Signature ,__logger__ : LoggerType,*args,**kwargs):
+    """Extracts the parameters of a function and traces the
+       call into logger (With trace level)
+    """
+    logger = __logger__
+    tid = unique_trace_id()
+    bound = sig.bind(*args,**kwargs)
+    non_defaults = list(bound.arguments.keys())
+    bound.apply_defaults()
+    log_str = f"Function call ({tid}):\n{func.__qualname__}("
+    for param,value in bound.arguments.items():
+        log_str += f"\n\t{param} " + ("" if param in non_defaults else "(default)") + f"= {value},"
+    log_str +=")"
+    logger.debug(log_str)
+    return tid
+
+def traced_func(f):
+    """
+    A tracing helper which logs function name and parameters so it
+    is easy to debug this module.
```

Comment:
I don't understand, is this to debug coverage_utils.py itself?

Reply:
Yes, as explained in the doc-string, if the logging level is set to debug, this will dump the parameters given to
any function that was annotated with this.

Reply:
Please remove it, it makes the file complicated.

I'm sure there are python debuggers and tracers.